### PR TITLE
[grammar-finder] Add new Core Package

### PR DESCRIPTION
**Pull request by confused-Techie**
Requires #907 

This **Draft** PR adds a new core package `grammar-finder` to Pulsar.

The decision to add this package is still under review, and until such time as it's agreed on, this PR will remain in draft status.

---

Grammar-Finder is a package that takes advantage of the new backend feature that allows searching of packages via the file extension that the grammar within the package supports.

This package comes with two ways to discover these packages:

* By manually triggering the Command Palette `grammar-finder:find-grammars-for-file` which will grab the extension of the current open file and search for any packages on the PPR that support said extension.
* `AutoFind` this feature is really the selling point of the package. It uses the new event introduced in #907 and if Pulsar is unable to find a suitable grammar for a file, opening the default Null Grammar, it'll automatically search for a community package that supports this file extension and create a notification for the user to view the list.

In implementing this package I tried to put as much room as possible for removing interactions the user doesn't want. Such as being able to use the command palette even if `AutoFind` is disabled to find a suitable package. 

Additionally, this package has a config that allows an array of file extensions that can be added , which it will ignore for any and all checks and never prompt for installable packages.

### CODE REVIEW

File: packages/grammar-finder/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Grammar-Finder
+
+Discover Community Package Grammars for your files.
+
+## AutoFind
+
+With 'AutoFind' enabled, when Pulsar fails to locate a grammar for the file you've just opened, defaulting to 'Plain Text', Grammar-Finder will automatically contact the Pulsar Package Registry in search of a Community Package that provides Syntax Highlighting for the file currently opened.
```
Comment (savetheclocktower) [inline packages/grammar-finder/README.md:7]:
```suggestion
With 'AutoFind' enabled, when Pulsar fails to locate a grammar for the file you've just opened, defaulting to 'Plain Text', `grammar-finer` will automatically contact the Pulsar Package Registry in search of a community package that provides syntax highlighting for the file currently opened.
```

File: packages/grammar-finder/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+If any packages are found you can easily view the whole list and install the one that looks best.
+
+When an 'AutoFind' notification appears you can quickly select:
+  * 'View Available Packages' to view the packages found.
+  * 'Disable Grammar-Finder for <ext>' to add this extension to the `ignoreExtList`.
+  * 'Disable AutoFind' to disable 'AutoFind' completely.
+
+## Command Palette
+
+Grammar-Finder adds `grammar-finder:find-grammars-for-file` to the Command Palette, so that at any time you can check if any Community Packages provide Syntax Highlighting for the file you are currently working in.
```
Comment (savetheclocktower) [inline packages/grammar-finder/README.md:18]:
```suggestion
`grammar-finder` adds `grammar-finder:find-grammars-for-file` to the Command Palette, so that at any time you can check if any community packages provide syntax highlighting for the file you are currently working in.
```

File: packages/grammar-finder/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+This is a great way to find alternative packages, or if you don't like the notifications of 'AutoFind', this can be used to locate packages on your terms.
+
+## Configuration
+
+### `autoFind`
+
+This can be enabled or disabled at will, effecting the 'AutoFind' feature availability.
```
Comment (savetheclocktower) [inline packages/grammar-finder/README.md:26]:
```suggestion
When enabled, `autoFind` will show a notification inviting you to install a suitable grammar for an unrecognized file type.
```

File: packages/grammar-finder/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+### `ignoreExtList`
+
+Any file extension can be added to this list to disable all checks for Community Packages. Ensure the extension is added without any preceding `.` for lookups to occur correctly.
```
Comment (savetheclocktower) [inline packages/grammar-finder/README.md:30]:
```suggestion
Any file extensions can be added to this list to disable all automatic checks for community packages for those file types. Choosing the “Disable `grammar-finder` for X” option on an `autoFind` notification will automatically add a given file extension to this list. This field should contain a comma-separated list of file extensions without any leading `.`s.
```

File: packages/grammar-finder/lib/main.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    }
+  }
+
+  async checkForGrammars(ext) {
+    this.superagent ??= require("superagent");
+
+    const res = await this.superagent
+      .get("https://api.pulsar-edit.dev/api/packages")
+      .set("User-Agent", "Pulsar.Grammar-Finder")
+      .query({ fileExtension: ext });
```
Comment (savetheclocktower) [inline packages/grammar-finder/lib/main.js:135]: I know you're familiar with `superagent`, but maybe we could use `fetch` here instead?

Just a low-stakes suggestion; this isn't a deal-breaker for me.

File: packages/grammar-finder/lib/main.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        buttons: [
+          {
+            text: "View Available Packages",
+            onDidClick: () => {
+              let packageListView = new PackageListView(packages);
+              packageListView.toggle();
+            }
+          },
+          {
+            text: `Disable Grammar-Finder for '${ext}'`,
```
Comment (savetheclocktower) [inline packages/grammar-finder/lib/main.js:161]: I wonder if we can state these actions more plainly in a way that doesn't make the user figure out what `grammar-finder` is. How about (e.g.)

<kbd>Don’t suggest packages for EXT files</kbd>

and

<kbd>Never suggest packages for unrecognized files</kbd>

?

File: packages/grammar-finder/package.json [hunk 7]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "grammar-finder",
+  "verison": "1.0.0",
+  "main": "./lib/main.js",
+  "description": "Discover Community Package Grammars for your files.",
```
Comment (savetheclocktower) [inline packages/grammar-finder/package.json:5]:
```suggestion
  "description": "Discover language packages for unrecognized files",
```

File: packages/grammar-finder/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
+# Grammar-Finder
+
+Discover Community Package Grammars for your files.
```
Comment (savetheclocktower) [inline packages/grammar-finder/README.md:3]: Discover language grammars for unrecognized files.

File: packages/grammar-finder/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+When an 'AutoFind' notification appears you can quickly select:
+  * 'View Available Packages' to view the packages found.
+  * 'Disable Grammar-Finder for <ext>' to add this extension to the `ignoreExtList`.
+  * 'Disable AutoFind' to disable 'AutoFind' completely.
+
+## Command Palette
+
+Grammar-Finder adds `grammar-finder:find-grammars-for-file` to the Command Palette, so that at any time you can check if any Community Packages provide Syntax Highlighting for the file you are currently working in.
+
+This is a great way to find alternative packages, or if you don't like the notifications of 'AutoFind', this can be used to locate packages on your terms.
```
Comment (savetheclocktower) [inline packages/grammar-finder/README.md:20]:
```suggestion
This makes it possible to find grammars for _recognized_ file types — or for unrecognized file types if you’ve disabled `autoFind`.
```

**Review by savetheclocktower — COMMENTED** [review]
Some suggestions for the prose. I’ll make broader suggestions in a follow-up comment.

**Comment by savetheclocktower** [comment]
A couple things:

1. I share the feeling expressed elsewhere that viewing a simple list of possible grammars to install might not give the user enough context. If I knew nothing about how you'd implemented this, and I just heard the feature described, I'd envision it linking to a search results page on `web.pulsar-edit.org`.

    There are certainly pros and cons to that approach, and keeping it all within the editor makes it a more streamlined experience. But suppose I open a `tex` file and am told that Pulsar can install a grammar for it. Great! [Which one do I pick?](https://web.pulsar-edit.dev/packages/search?fileExtension=tex) If I had to pick quickly, probably the one with more downloads — but the top two candidates are pretty close to one another in that respect, and I'd probably want to dig into each one and see which has more features, which was more recently updated, et cetera.

    I don't want to make an ultimatum here because I'm certainly _not_ the intended user of this feature. Just mentioning it in case a lot of other people have the same feeling.

2. Let's make sure that, when tests are added, they cover these edge cases. I'm not saying I know exactly what the package should do in each of these situations, but let's at least brainstorm about it ahead of time:

    * What happens when the file type actually _warrants_ a plain text grammar? Like a `txt` file, or a `README` with no file extension? Those two might be easy to guard against, but I'll bet there are plenty more. At the very least, those two file extensions should be listed in `ignoreExtList` by default.
    * This won't appear when the user creates a new untitled document, right? I gather that the lack of a file type would preclude the notification?
    * What happens when the user already has a package that can understand the given file type, but that package is disabled? Do we try to detect that scenario?
    * What happens if the user hits <kbd>Esc</kbd> or clicks the close button? As in “I’m not choosing any of these options explicitly; I’m just busy and would prefer to think about it later”? (I think we can prevent the notification from being dismissed in these manners, but should we?) Do we take that to mean “ask again literally the next time they open this kind of file”? Or do we try to set a flag not to ask them again… when? For 24 hours? Until the window is reloaded?

**Comment by confused-Techie** [comment]
@savetheclocktower I appreciate the extremely well thought out review and questions here, and I'll go ahead and address what I can think of right away, but otherwise will likely need to think a bit on a few of these.

For your questions on behaviors I want to quickly draw the distinction between the Command Palette invocation of this package and the `AutoFind` feature. When this package is triggered via the Command Palette it will always show community packages unless it finds none, or that extension is listed as an ignored extension. But for the rest of my answers I'll say them strictly from the perspective of the `AutoFind` feature.

## When `text.plain` is the best grammar 

This package checks strictly for `text.plain.null-grammar`. Which as far as I can tell is only used when Pulsar can't find the right grammar.

For example when I open `test.txt` the grammar scope name chosen is `text.plain`. So this AutoFind wouldn't do anything, as a proper grammar was selected.

So as long as the actually plain text grammar is being used, then this package won't act on it. Of course there may be a situation where plain text is the best and Pulsar also happens to be falling back to a null grammar, but unless we can come up with a common list of extensions where this happens, that may be somewhat impossible to protect against. And somewhat against the purpose of the AutoFind feature, which is intended to only act when Pulsar can't find a grammar.

But to continue, your example of a file named `README` is a good one. Since Pulsar falls back to `text.plain.null-grammar`. Which makes does cause this package to act. But in this case, since there's no community packages that offer support of this extension (Which is detected as `README`), nothing happens. Since I did really want to make this package as unobtrusive as possible for any automatic features. If the package errors out (in a known way) or otherwise can't find any packages, it just does nothing. Not bothering the user at all.

So the second part of the answer to when text is really the best, I'd like to think that is a self solving problem. If text is truly the best, no community packages will exist to support it, so that this package will end up failing to find any suitable community packages and end up doing nothing. Sure that all comes at the cost of a few CPU cycles wasted, but is nice in theory.

## Blank New Document

You are right, in this case no extension can be identified, so it's unable to find any packages at all. Simple staying silent and doing nothing.

## If a suitable package is currently disabled

As of now we do absolutely nothing to detect this situation. Since the only sources of information for this package come from:
* The current buffer's file extension
* The current scope of the selected grammar
* The API results from the backend

Nothing beyond that is currently considered, but if we thought it important to do nothing in this situation we could certainly look at implementing it. Since I'm all for making this package as unobtrusive as possible.

## Hitting `Esc`

If `esc` is hit while viewing the list of packages, only that modal goes away. Leaving the notification up until it's closed or `esc` is hit again. If the notification is closed we currently do nothing to prevent asking again.

But adding this behavior is honestly a really good idea. I'd personally say we cache extension dismissals until the current window is closed. We could just store that as a local variable of the package, and let it get destroyed naturally. Then of course if the user wants a more non-temporal way to dismiss that extension, they could always add it to the extension ignore list.

---

But for the concerns of making informed decisions about the best package, that seems to be the most obvious thing lacking so far, and will absolutely take some time to figure out how best to address it. Whether that means making that information available in this view, or passing off the actual request to `settings-view` remains to be seen.

Again thanks for taking the time to look at this one

File: packages/grammar-finder/README.md (hunk 4 above)
Comment (confused-Techie) [reply packages/grammar-finder/README.md:30]: I especially had trouble writing this one up, thanks for clarifying it tons

File: packages/grammar-finder/README.md (hunk 8 above)
Comment (confused-Techie) [inline packages/grammar-finder/README.md:3]:
```suggestion
Discover language grammars for unrecognized files.
```

File: packages/grammar-finder/lib/main.js (hunk 6 above)
Comment (confused-Techie) [reply packages/grammar-finder/lib/main.js:161]: Yeah these are really good.
I hadn't considered the effort needed to place some random core package, but often times a user may not even know it exists when they see this prompt. So that's a really good idea to make the prompts as digestible as possible without any internal knowledge.

Comment (confused-Techie) [reply packages/grammar-finder/lib/main.js]: I've gone ahead and implemented your awesome suggestions

**Comment by confused-Techie** [comment]
Alright, added your suggestions as well as made one small change:

I like what you said about trying not to prompt twice, so I've added a very simple implementation for this.

In a local non-cached variable we track every extension AutoFind prompts for. Then we don't prompt if the extension is on that list.

So that in the event the user opens an unrecognized file then closes it and opens again, they won't get two of the same notification.

Or if they open multiple of the same kind of file they won't get a prompt for every single one. 

Hopefully this can cull the lowest hanging fruit to this problem

File: packages/grammar-finder/lib/main.js (hunk 5 above)
Comment (confused-Techie) [reply packages/grammar-finder/lib/main.js:146]: For such a simple request like this that's a very good idea. Plus it'd cut our dependencies in half

File: packages/grammar-finder/README.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
+# Grammar-Finder
+
+Discover language grammars for unrecognized files.
+
+## AutoFind
+
+With 'AutoFind' enabled, when Pulsar fails to locate a grammar for the file you've just opened, defaulting to 'Plain Text', `grammar-finer` will automatically contact the Pulsar Package Registry in search of a community package that provides syntax highlighting for the file currently opened.
```
Comment (Daeraxa) [inline packages/grammar-finder/README.md:7]:
```suggestion
With 'AutoFind' enabled, when Pulsar fails to locate a grammar for the file you've just opened, defaulting to 'Plain Text', `grammar-finder` will automatically contact the Pulsar Package Registry in search of a community package that provides syntax highlighting for the file currently opened.
```

**Comment by Daeraxa** [comment]
On the topic of "known plaintext files" how does/should it behave with files that have no extension - common for Linux config files.

There is also one format I can imagine could both be annoying and useful at the same time which would be for `.log`. The chance of there being a correct grammar for your specific `.log` file is rather low but there are already three packages out there that do supply for that extension - just not anything I personally would have much use for.

**Comment by confused-Techie** [comment]
@Daeraxa For files that have no extension, it will end up looking at the name of the file for the extension.

This is because of how NodeJS's `path.parse()` method behaves, I slightly disagree with it.

Take for example the following:

* `.nvm` => `{ name: "nvm", ext: "" }`

Here I'd personally say that the extension is `nvm` and the file has no actual name, but NodeJS disagrees. But even more existing packages think the same way here, for example the package that provides syntax highlighting for GitIgnore files declares the extension as `gitignore`. So if we don't get an extension from `path.parse()` we fallback to using the name just in case this is happening.

Which will also mean for files that legitimately have no extension like `README` the name will be used instead. Which if memory serves will likely be the same way it's handled within Pulsar itself, so hopefully that stays consistent.

As for any specific examples like `log` that could arguably be a default on the ignored extension list, since usually they are rather specific in terms of a usable grammar, but it's hard to say without taking a deeper look at what currently exists.

File: packages/grammar-finder/lib/main.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+const PackageListView = require("./package-list-view.js");
+
+class GrammarFinder {
+  activate() {
+
+    // This local variable is intended to act as 'session' storage, or editing
+    // session storage. Where the next time the editor is opened it's info is gone
+    this.promptedForExt = [];
+
+    atom.grammars.emitter.on("did-auto-assign-grammar", async (data) => {
```
Comment (mauricioszabo) [inline packages/grammar-finder/lib/main.js:12]: This relates to my comment: https://github.com/pulsar-edit/pulsar/pull/907#issuecomment-2085527555

Basically, we usually have an API like `onDid...` to add callbacks. I think it's a good idea to add one to the other PR, and use it here.

In any way - we should _never_ add a callback without a cleanup, so we might want to add a `CompositeDisposable` here and dispose it when deactivating package

File: packages/grammar-finder/lib/main.js [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+        // we have been told to ignore this ext
+        return;
+      }
+
+      if (this.promptedForExt.includes(ext)) {
+        // If we have already prompted for this extension in this editing session
+        return;
+      }
+
+      const packages = await this.checkForGrammars(ext);
```
Comment (mauricioszabo) [inline packages/grammar-finder/lib/main.js:38]: I don't think you can use `await` here, right? We need `activate` to be async, or move this to a different method...

File: packages/grammar-finder/lib/main.js [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+          if (packages.length === 0) {
+            // No packages were found that support this grammar
+            // since manuall invoked we may want to notify
+            atom.notifications.addInfo(`Unable to locate any Grammars for '${ext}'.`, { dismissable: true });
+            return;
+          }
+
+          // Lets notify the user about the found packages
+          this.notify(packages, ext, `'${packages.length}' Installable Grammars are available for '${ext}'.`);
+        }
```
Comment (mauricioszabo) [inline packages/grammar-finder/lib/main.js:98]: Can we move all of this to a method, so it's easier to read?

File: packages/grammar-finder/lib/main.js [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+      })
+    );
+  }
+
+  deactivate() {
+    this.superagent = null;
+  }
+
+  inspectAssignment(data, opts = {}) {
+    console.log(`grammar-finder.inspectAssignment(${data.grammar.scopeName}, ${data.buffer.getPath()})`);
```
Comment (mauricioszabo) [inline packages/grammar-finder/lib/main.js:108]: Can we only log on dev mode? Seems like a "debug info" here.

File: packages/grammar-finder/lib/main.js [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+      // in Pulsar, the same isn't true.
+      let ext = false;
+
+      if (typeof parsed.ext === "string" && parsed.ext.length > 0) {
+        ext = parsed.ext;
+      } else if (typeof parsed.name === "string" && parsed.name.length > 0) {
+        ext = parsed.name;
+      }
+
+      console.log(`File: ${filePath} - Ext: ${ext}`);
```
Comment (mauricioszabo) [inline packages/grammar-finder/lib/main.js:132]: Same here about a dev/debug mode

File: packages/grammar-finder/lib/main.js [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+
+      return ext;
+    } else {
+      return false;
+    }
+  }
+
+  async checkForGrammars(ext) {
+    this.superagent ??= require("superagent");
```
Comment (mauricioszabo) [inline packages/grammar-finder/lib/main.js:141]: What is superagent?

File: packages/grammar-finder/styles/grammar-finder.less [hunk 17]
```diff
@@ -0,0 +1,90 @@
+@import "ui-variables";
+
+@package-card-background-color: lighten(@tool-panel-background-color, 8%);
+
+.list-group .package-card.selected::before {
```
Comment (mauricioszabo) [inline packages/grammar-finder/styles/grammar-finder.less:5]: Because of the way CSS works on Pulsar, it's always a good idea to make the "core" selector the same as the package name. So, instead of `.package-card`, use something like `.grammar-finder`, or even `grammar-finder.package-card`, WDYT?

**Review by mauricioszabo — CHANGES_REQUESTED** [review]
